cmake/Version.cmake

```diff
@@ -8,20 +8,100 @@
 
 string(TIMESTAMP current_year "%Y")
 set(O3DE_COPYRIGHT_YEAR ${current_year} CACHE STRING "Open 3D Engine's copyright year")
-set(O3DE_VERSION_STRING "0.0.0" CACHE STRING "Open 3D Engine's version")
-set(O3DE_DISPLAY_VERSION_STRING "00.00" CACHE STRING "Open 3D Engine's display version")
-set(O3DE_BUILD_VERSION 0 CACHE STRING "Open 3D Engine's build number")
-set(O3DE_ENGINE_NAME "o3de" CACHE STRING "Open 3D Engine's engine name")
-
-# Optional environment overrides
-if(NOT "$ENV{O3DE_VERSION}" STREQUAL "")
-    set(O3DE_VERSION_STRING "$ENV{O3DE_VERSION}")
-endif()
-
-if(NOT "$ENV{O3DE_DISPLAY_VERSION}" STREQUAL "")
-    set(O3DE_DISPLAY_VERSION_STRING "$ENV{O3DE_DISPLAY_VERSION}")
-endif()
-
-if(NOT "$ENV{O3DE_BUILD_VERSION}" STREQUAL "")
-    set(O3DE_BUILD_VERSION "$ENV{O3DE_BUILD_VERSION}")
-endif()
+
+# avoid reading engine.json multiple times 
+ly_file_read("${LY_ROOT_FOLDER}/engine.json" tmp_json_data)
+set_property(GLOBAL PROPERTY O3DE_ENGINE_JSON_DATA ${tmp_json_data})
+unset(tmp_json_data)
+
+#! o3de_read_engine_default: Read a field from engine.json or use the default if not found
+# \arg:output_value - name of output variable to set 
+# \arg:key - name of field in engine.json 
+# \arg:default_value - value to use if neither environment var or engine.json value found 
+macro(o3de_read_engine_default output_value key default_value)
+    get_property(engine_json_data GLOBAL PROPERTY O3DE_ENGINE_JSON_DATA)
+    string(JSON tmp_value ERROR_VARIABLE manifest_json_error GET ${engine_json_data} ${key})
+
+    # unset engine_json_data because we're in a macro
+    unset(${engine_json_data})
+
+    if(manifest_json_error)
+        message(WARNING "Failed to read ${key} from \"${LY_ROOT_FOLDER}/engine.json\" : ${manifest_json_error}")
+        set(${tmp_value} ${default_value})
+    endif()
+
+    ly_set(${output_value} ${tmp_value})
+    set_property(GLOBAL PROPERTY ${output_value} ${tmp_value})
+
+    # unset tmp_value because we're in a macro
+    unset(${tmp_value})
+endmacro()
+
+#! o3de_set_install_default: Sets the output_value to one from the environment
+#  or the default, if the environment variable is empty
+# \arg:output_value - name of output variable to set 
+# \arg:env_var - name of environment variable to use 
+# \arg:default_value - value to use if environment var is empty
+macro(o3de_set_install_default output_value env_var default_value description)
+    set(${output_value} ${default_value} CACHE STRING ${description})
+    if(NOT ${env_var} STREQUAL "")
+        set(${output_value} ${env_var})
+    endif()
+    string(CONFIGURE ${${output_value}} ${output_value} @ONLY)
+endmacro()
```

Comment:
This would be a good to move to `LySet.cmake` since this is general purpose function.
I don't have a short name that explains all what the function does, but how about
`o3de_set_in_env_or_cache_and_substitute_var_refs`

Reply:
would you recommend I use `ly_set` instead of `set` on line 48?

Reply:
Not really..., Because that would result in different expectations which scope the variable is set in.
For example this function being called in the Version.cmake file would result in the values being set in the engine(`<engine-root>`) and project(`<project-root>`) scopes in a project-centric workflow and only the engine scope(`<engine-root>` in an engine-centric flow.

What this means is that in an engine-centric with the path to the AutomatedTesting project set via `LY_PROJECTS=./AutomatedTesting`, the AutomatedTesting project will not have access to the variable set by this function.
However if the AutomatedTesting project was run in a project workflow(i.e you run CMake configure from inside the AutomatedTestingProject, then it would have access to that variable).

To avoid that inconsistency, I think it is better to just set the value in the current scope and then let the caller deal with setting it in any other scopes.